File: plugins/collector/event/apps/container/kubelet.go
package container

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Kubelet struct {
	version string
	reg     *regexp.Regexp
}

func (Kubelet) Name() string { return "kubelet" }

func (Kubelet) Type() string { return "container" }

func (d Kubelet) Version() string { return d.version }

func (Kubelet) Match(p *process.Process) bool { return p.Name == "kubelet" }

func (n *Kubelet) Run(p *process.Process) (m map[string]string, err error) {
	if n.reg == nil {
		n.reg = regexp.MustCompile(`v(\d+\.)+\d+\S+`)
	}
	result, err := apps.Execute(p, "--version")
	str := n.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	n.version = strings.TrimPrefix(str, "v")
	return
}

func init() { apps.Regist(&Kubelet{}) }

File: plugins/collector/src/event/apps/container/kubelet.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Kubelet {
    version: String,
}

impl Kubelet {
    pub fn new() -> Self { Kubelet { version: String::new() } }
}

impl IApp for Kubelet {
    fn name(&self)     -> &'static str { "kubelet" }
    fn app_type(&self) -> &'static str { "container" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "kubelet" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "Kubernetes v1.25.2"
        if let Some(v) = out.split_whitespace().find(|s| s.starts_with('v')) {
            self.version = v.trim_start_matches('v').to_owned();
        } else {
            self.version = find_version(&out);
        }
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
